### PR TITLE
Add support for three or more chord keyboard shortcuts

### LINKED ISSUE #6966: enable keychords of 3 or more steps
- VSCode Version: 1.1.1
- OS Version: Windows7, x64

Steps to Reproduce:
1. Create a key-chord of 3 steps: "ctrl+k ctrl+c d" and bind it to a command
2. Try the new key-chord
3. Observe no action

Expected: 3 steps key-chords should work.

### PR DESCRIPTION
**Pull request by dyedgreen**
Closes #6966.

Allow for three or more steps in keyboard shortcut cords.

This change only allows for such chords to be configured by the user in the key bindings JSON (by adding multiple steps with spaces; this was already parsed correctly, but the key binding machinery just ignored any keys past the first two).

The interactive keybinding recording widget is also updated to allow for unlimited length chord sequences.

Roughly, this is done by addressing the existing `TODO@chords` comments in the code.

### CODE REVIEW

**Comment by alexdima** [comment]
@dyedgreen Thank you for your PR. In order for me to be able to review it, please accept the Contributor License Agreement  (see above comment)

**Comment by dyedgreen** [comment]
Thanks fit getting back to me so quickly @alexdima!

I’m checking with my employer to see if I need to include them here / should accept purely myself and will accept the CLA here as soon as they get back to me 😅

**Comment by dyedgreen** [comment]
@microsoft-github-policy-service agree company="Jane Street"

File: src/vs/platform/keybinding/common/keybindingResolver.ts
Comment (ulugbekna) [inline src/vs/platform/keybinding/common/keybindingResolver.ts:161]: hm, I'm not sure whether difference in length between `item.chords` and `conflict.chords` would be a "divergence" here

If `conflict` is `cmd+k cmd+k` and `item` is `cmd+k cmd+k cmd+k`, `conflict` would not be removed from lookup, but also unclear to me whether a user would want it to be

**Review by ulugbekna — DISMISSED** [review]
Thanks for the PR @dyedgreen  (a fellow OCaml'er, I guess? 🐫)

The PR looks good, a couple of inline comments.

File: src/vs/platform/keybinding/common/keybindingResolver.ts
Comment (dyedgreen) [reply src/vs/platform/keybinding/common/keybindingResolver.ts:165]: Yes, agreed!

File: src/vs/platform/keybinding/common/keybindingResolver.ts
Comment (dyedgreen) [reply src/vs/platform/keybinding/common/keybindingResolver.ts:161]: I think the logic here makes sense, although it's a little subtle when the keybindings have differing lengths. (I renamed the variable to `chordsFullyOverlap`, which hopefully makes it a bit more clear?)

Essentially, if the keybinding overlaps completely (as in the shorter keybinding is fully contained within the longer one), they logically conflict, since when the chord sequence is pressed, it triggers the shorter keybinding, then exits the chord mode. In this case, the shorter keybinding completely shadows the longer one. (And I think it makes sense to flag this in the UI + treat it in the same way as other conflicts, with the later defined binding taking precedence).

I will also leave a comment to that effect here? (Feel free to edit the comment if you think it's confusing)!

File: src/vs/platform/keybinding/test/common/keybindingsTestUtils.ts
Comment (dyedgreen) [reply src/vs/platform/keybinding/test/common/keybindingsTestUtils.ts:13]: I think this is fine (?)

As far as I know `x === 0` is equivalent to `typeof x == "number" && x == 0`, and since `typeof [] === "object"` we'd get the same check. (My understanding / assumption was that TS considers `0` to be a member of the type `number | number[]`, but I don't really know TS in too much detail so that's just a semi-educated guess)

File: src/vs/platform/keybinding/common/keybindingResolver.ts
Comment (dyedgreen) [reply src/vs/platform/keybinding/common/keybindingResolver.ts:164]: E.g. for your example: `cmd+k cmd+k cmd+k` and `cmd+k cmd+k`, we'd find that `shorterChordIsPrefix = true`, and so we won't fall out of the loop iteration via continue, perform the `whenIsEntirelyIncluded` check, and treat the keybinding as a conflict.

(At least that's my understanding of the code as written, please correct me if I'm wrong about how this interacts with the rest of the system 😅)

**Comment by dyedgreen** [comment]
Thanks for the review @ulugbekna! I've made a few minor tweaks based on your comments, let me if they make sense.

And yes, we famously write even our [hardware](https://github.com/janestreet/hardcaml) in OCaml 😄

File: src/vs/platform/keybinding/common/keybindingResolver.ts
Comment (ulugbekna) [reply src/vs/platform/keybinding/common/keybindingResolver.ts:163]: Makes sense, thanks!

**Review by ulugbekna — DISMISSED** [review]

**Comment by dyedgreen** [comment]
Just to check, is there anything else required from my side for this to be merged?

**Comment by ulugbekna** [comment]
This is awaiting @alexdima 's review. He's very occupied right now, but we haven't forgotten about the PR. Thanks for patience with this

**Comment by ulugbekna** [comment]
@dyedgreen btw, would you be interested in implementing support for multi-chord keybindings in the keybindings editor? It would become complete and more likely to be shipped.

**Comment by dyedgreen** [comment]
I can take a look at how hard that would be 👍

**Comment by dyedgreen** [comment]
@ulugbekna I've updated the PR to also allow unlimited chords in the keybinding recording widgets.

This means if you miss-click while recording, you have to abort (escape) and retry; previously it would let you cycle and try again although I personally found that confusing.

**Review by ulugbekna — CHANGES_REQUESTED** [review]
First, big thanks for being active with the PR and great work. The changes look good. I see two issues: 

1. Search by keybinding prefix works only for 2 chords, ie: 
    - `cmd+k` finds keybindings that have their first chord `cmd+k`
    - `cmd+k cmd+k` finds keybindings that have their 1st and 2nd chord `cmd+k cmd+k`
    - but `cmd+k cmd+k cmd+k` doesn’t find the keybinding `cmd+k cmd+k cmd+k cmd+i` - user has to insert the exact keybinding to find itI think it would be great to fix before merging. 

I think this needs to be fixed before merging. 

2. In the define keybinding widget, we don't resize on overflow. I don't think people would be defining KBs with > 5 chords, but this remains an issue. 

image

**Comment by ulugbekna** [comment]
@sandy081 it would be great to have a review from you for the keybinding widget changes

**Comment by ulugbekna** [comment]
Upd: I'll make a commit to this PR that restricts the UI to 2 chords, so that we don't burden you with more work and merge the PR sooner. We'll address the UI changes in another PR.

**Review by alexdima — CHANGES_REQUESTED** [review]
This looks great! I've pushed a change to limit the chord count to 2 in the widget for now. We should improve this in the widget itself via separate PRs. However, there is one problem where modifier only keybindings no longer work e.g.:
```
    {
      "key": "shift shift",
      "command": "actions.find"
    }
```

File: src/vs/platform/keybinding/common/abstractKeybindingService.ts
Comment (dyedgreen) [reply src/vs/platform/keybinding/common/abstractKeybindingService.ts:278]: I think it might not. I’ll take a look at how we can make this work again 👀

Comment (dyedgreen) [reply src/vs/platform/keybinding/common/abstractKeybindingService.ts]: I pushed some changes to restore the previous behavior. The way the single modifier chords are set-up, they can't be chorded to arbitrary length, but I feel like that's okay? (It fully restores the old behavior, while allowing arbitrary length chords for regular keys.)

**Review by alexdima — APPROVED** [review]
:+1: Thank you!

**Review by ulugbekna — APPROVED** [review]
great work, thanks heaps! merging :-)

**Comment by anddam** [comment]
Thanks @dyedgreen  and everybody involved for this especially.

When can end users expect this commit to land in a stable MS release?

**Pull request merged**
